weave/trace_server/clickhouse_trace_server_batched.py

```diff
@@ -1390,3 +1464,350 @@ def find_all_descendants(root_ids: typing.List[str]) -> typing.Set[str]:
+    query = f"SELECT * FROM some_table WHERE some_column = {{{param_name}:String}}"
+    parameters = pb.get_params()
+    # Execute the query with the parameters
+    ```
+
+    With queries that have many construction phases, it is recommended to use the
+    same ParamBuilder instance to ensure that the parameter names are unique across
+    the query.
+    """
+
+    def __init__(self, prefix: typing.Optional[str] = None):
+        global param_builder_count
+        param_builder_count += 1
+        self._params: typing.Dict[str, typing.Any] = {}
+        self._prefix = (prefix or f"pb_{param_builder_count}") + "_"
+
+    def add_param(self, param_value: typing.Any) -> str:
+        param_name = self._prefix + str(len(self._params))
+        self._params[param_name] = param_value
+        return param_name
+
+    def get_params(self) -> typing.Dict[str, typing.Any]:
+        return {**self._params}
+
+
+def _python_value_to_ch_type(value: typing.Any) -> str:
+    """Helper function to convert python types to clickhouse types."""
+    if isinstance(value, str):
+        return "String"
+    elif isinstance(value, int):
+        return "UInt64"
+    elif isinstance(value, float):
+        return "Float64"
+    elif isinstance(value, bool):
+        return "UInt8"
+    elif value is None:
+        return "Nullable(String)"
+    else:
+        raise ValueError(f"Unknown value type: {value}")
+
+
+def _param_slot(param_name: str, param_type: str) -> str:
+    """Helper function to create a parameter slot for a clickhouse query."""
+    return f"{{{param_name}:{param_type}}}"
+
+
+def _quote_json_path(path: str) -> str:
+    """Helper function to quote a json path for use in a clickhouse query. Moreover,
+    this converts index operations from dot notation (conforms to Mongo) to bracket
+    notation (required by clickhouse)"""
+    parts = path.split(".")
+    parts_final = []
+    for part in parts:
+        try:
+            int(part)
+            parts_final.append("[" + part + "]")
+        except ValueError:
+            parts_final.append('."' + part + '"')
+    return "$" + "".join(parts_final)
+
+
+def _is_dynamic_field(field: str) -> bool:
+    """Dynamic fields are fields that are arbitrary values produced by the user."""
+    return (
+        field in ["inputs", "output", "attributes", "summary"]
+        or field.startswith("inputs.")
+        or field.startswith("output.")
+        or field.startswith("attributes.")
+        or field.startswith("summary.")
+    )
+
+
+def _transform_external_calls_field_to_internal_calls_field(
+    field: str,
+    cast: typing.Optional[str] = None,
+    param_builder: typing.Optional[ParamBuilder] = None,
+) -> tuple[str, ParamBuilder, set[str]]:
+    """Transforms a request for a dot-notation field to a clickhouse field."""
+    param_builder = param_builder or ParamBuilder()
+    raw_fields_used = set()
+    json_path = None
+    if field == "inputs" or field.startswith("inputs."):
```

Comment:
Could consider using a loop since this is on the repetitive side.

Reply:
fixed in next commit